sandbox/cgroup/kill.go

```diff
@@ -0,0 +1,122 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+// KillSnapProcesses sends a signal to all the processes belonging to
+// a given snap.
+//
+// A correct implementation is picked depending on cgroup v1 or v2 use in the
+// system. When cgroup v1 is detected, the call will directly act on the freezer
+// group created when a snap process was started, while with v2 the call will
+// act on all tracking groups of a snap.
+//
+// XXX: It is important to note that the implementation for v2 is racy because
+// processes can be killed externally even when the cgroup is frozen and have
+// their pid reused. Also, v2 freezing support is a no-op on kernels before
+// 5.2 which means new processes can keep getting spawned while killing
+// pids read earlier.
+var KillSnapProcesses = killSnapProcessesImplV1
+
+func pickKillV1Impl() {
+	KillSnapProcesses = killSnapProcessesImplV1
+}
+
+func pickKillV2Impl() {
+	KillSnapProcesses = killSnapProcessesImplV2
+}
+
+var syscallKill = syscall.Kill
+
+// killProcessesInCgroup sends signal to all the processes belonging to
+// passed cgroup directory.
+//
+// The caller is responsible for making sure that pids are not reused
+// after reading `cgroup.procs` to avoid TOCTOU.
+func killProcessesInCgroup(dir string, signal syscall.Signal) error {
+	pids, err := pidsInFile(filepath.Join(dir, "cgroup.procs"))
+	if err != nil {
+		return err
+	}
+
+	var firstErr error
+	for _, pid := range pids {
+		err := syscallKill(pid, signal)
+		if err != nil && err != syscall.ESRCH && firstErr == nil {
+			firstErr = err
+		}
+	}
+
+	return firstErr
+}
+
+func killSnapProcessesImplV1(ctx context.Context, snapName string, signal syscall.Signal) error {
+	if err := freezeSnapProcessesImplV1(ctx, snapName); err != nil {
+		return err
+	}
+	defer thawSnapProcessesImplV1(snapName)
+
+	return killProcessesInCgroup(filepath.Join(freezerCgroupV1Dir, fmt.Sprintf("snap.%s", snapName)), signal)
+}
+
+// XXX: killSnapProcessesImplV2 is racy to varying degrees depending on the kernel
+// version.
+//
+//  1. Cgroup v2 freezer was only available since Linux 5.2 so freezing is a no-op before 5.2 which allows processes to keep forking.
+//  2. Freezing does not put processes in an uninterruptable sleep unlike v1, so they can be killed externally and have their pid reused.
+//  3. `cgroup.kill` was introduced in Linux 5.14 and solves the above issues as it kills the cgroup processes atomically.
+func killSnapProcessesImplV2(ctx context.Context, snapName string, signal syscall.Signal) error {
+	action := func(dir string) error {
+		if signal == syscall.SIGKILL {
```

Comment:
Do we ever send other signals? Perhaps we can just require the signal to be SIGKILL and drop the logic?

Reply:
no, we currently don't. I can drop it for now and if we ever need other signals we can add support later

Reply:
Yeah, let's reduce this. Simple is good.